src/pending_connection/connect.rs

```diff
@@ -1,208 +1,261 @@
 use std::net::{IpAddr, SocketAddr};
-use std::time::{Duration, Instant};
+use std::time::Duration;
 
 use failure::Error;
-
 use futures::prelude::*;
 use futures::select;
-
-use log::{debug, info, warn};
 use tokio::time::interval;
+use tokio::time::Instant;
 
-use crate::packet::{
-    ControlPacket, ControlTypes, HandshakeControlInfo, HandshakeVSInfo, Packet, ShakeType,
-    SocketType, SrtControlPacket, SrtHandshake, SrtShakeFlags,
-};
+use log::warn;
+
+use crate::packet::*;
 use crate::util::get_packet;
-use crate::{Connection, ConnectionSettings, SocketID, SrtVersion};
+use crate::{Connection, ConnectionSettings, DataPacket, SeqNumber, SocketID, SrtVersion};
 
-pub async fn connect<T>(
-    sock: &mut T,
+use ConnectError::*;
+use ConnectState::*;
+
+pub struct ConnectConfiguration {
```

Comment:
The choice of having this `pub` is interesting--it's not really usable from an outside module, as there is no way to construct it. Maybe a good way to do this would be to have a integration test that uses this state machine without the `async connect` function